### PR TITLE
New Statistics page

### PR DESCRIPTION
**Pull request by dracos**
Reopen of #55, sorry.

screenshot_2018-08-23 statistics

### CODE REVIEW

**Comment by struan** [comment]
Now with graphs:

screenshot_2018-09-07 statistics

File: templates/web/smidsy/reports/index.html [hunk 1]
```diff
@@ -1,55 +1,142 @@
...
-<th class="data">[% loc('Older<br>near misses') %]</th>
-</tr>
-</thead>
+<div class="dashboard-row">
+    <div class="dashboard-item dashboard-item--12">
+        <div class="labelled-line-chart">
+            <canvas id="chart-all-reports" width="600" height="300"
+                data-labels="[&quot;[% problem_periods.join('&quot;,&quot;') %]&quot;]"
+                data-values-reports="[[% problems_reported_by_period.join(',') %]]"
+                data-values-fixed="[[% problems_fixed_by_period.join(',') %]]"
```
Comment (dracos) [inline templates/web/smidsy/reports/index.html]: Anyway we can not show the fixed line if it's always just 0s along the bottom?

File: templates/web/smidsy/reports/index.html [hunk 2]
```diff
@@ -1,55 +1,142 @@
...
+                <option value="">Collideoscope &amp; STATS19</option>
+                <option value="collideoscope"[% ' selected' IF sources == 'collideoscope' %]>Collideoscope only</option>
+                <option value="stats19"[% ' selected' IF sources == 'stats19' %]>STATS19 only</option>
+            </select>
+        </p>
+        <p>
+            <label for="transport">[% loc('Transport') %]</label>
+            <select class="form-control" id="transport" name="transport">
+                <option value="">All</option>
+                [% FOR p IN participant_labels.keys %]
```
Comment (dracos) [inline templates/web/smidsy/reports/index.html]: Perhaps use `pairs` here (and then `.key` and `.value` below) so it's sorted? Plus tiny bit clearer.

File: perllib/FixMyStreet/Cobrand/Smidsy.pm [hunk 3]
```diff
@@ -416,6 +416,307 @@ sub reports_hook_restrict_bodies_list {
...
+
+        my @bodies = $c->model('DB::Body')->active->translated->with_area_count->all_sorted;
+        @bodies = @{$c->cobrand->call_hook('reports_hook_restrict_bodies_list', \@bodies) || \@bodies };
+        $c->stash->{bodies} = \@bodies;
+
+        $c->stash->{start_date} = DateTime::Format::Strptime->new( pattern => "%H:%M:%S %d:%m:%Y" )->format_datetime($start);
+        $c->stash->{end_date} = DateTime::Format::Strptime->new( pattern => "%H:%M:%S %d:%m:%Y" )->format_datetime($end);
+        $c->stash->{severities} = \@severities;
+        $c->stash->{reports_by_severity} = \%reports_by_severity;
+        $c->stash->{participants} = [ keys %reports_by_participants ];
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Smidsy.pm]: Can we sort this, either alphabetically or in order of 'severity' or something?

File: perllib/FixMyStreet/Cobrand/Smidsy.pm [hunk 4]
```diff
@@ -416,6 +416,307 @@ sub reports_hook_restrict_bodies_list {
     }
 }
 
+sub loop_period {
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Smidsy.pm]: As this is a copy (slightly extended for generic `$end`) of the method in UpdateAllReports, could we put it in core somewhere both can use it?

File: perllib/FixMyStreet/Cobrand/Smidsy.pm [hunk 5]
```diff
@@ -416,6 +416,307 @@ sub reports_hook_restrict_bodies_list {
...
+    my $children = $body->first_area_children;
+    $self->{c}->stash->{area_body_name} = $body->name;
+    return [ values %$children ];
+}
+
+sub report_page_data {
+    my $self = shift;
+    my $c = $self->{c};
+
+    if (my $body = $c->get_param('body')) {
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Smidsy.pm]: This block is also in core, but after this fn is called. Could it be moved in core to be before this block is called, so it's not needed to be repeated here?

File: templates/web/smidsy/reports/index.html [hunk 6]
```diff
@@ -1,55 +1,142 @@
...
+            </tr>
+          [% END %]
+        </table>
+    </div>
+    <div class="dashboard-item dashboard-item--6">
+        <h2 class="dashboard-subheading">[% loc('Reports by transport mode') %]</h2>
+        <table class="dashboard-ranking-table js-make-bar-chart">
+          [% FOR participant IN participants %]
+            <tr>
+                <th scope="row">[% participant %]</th>
```
Comment (dracos) [inline templates/web/smidsy/reports/index.html]: Think this should be the value not the key? / Or use _labels

File: perllib/FixMyStreet/Cobrand/Smidsy.pm [hunk 7]
```diff
@@ -416,6 +416,307 @@ sub reports_hook_restrict_bodies_list {
...
+    );
+
+    $c->stash->{participant_labels} = \%participants;
+
+    if ( my $transport = $c->get_param('transport') ) {
+        if ( $participants{$transport} ) {
+            $c->stash->{transport} = $transport;
+            my $rs = $c->stash->{problems_rs};
+            my $filter = "SIMILAR TO '%T12:participants,T\\d+:$transport,%'";
+            $c->stash->{problems_rs} = $rs->search( { extra => \$filter } );
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Smidsy.pm]: This is a SQL injection. You need to use placeholders as otherwise it is not escaped. See https://metacpan.org/pod/SQL::Abstract#Literal-SQL-with-placeholders-and-bind-values-(subqueries)

File: templates/web/smidsy/reports/index.html [hunk 8]
```diff
@@ -1,55 +1,142 @@
...
-      [% IF report_filter == 'city' %]checked[% END %]> 
-      Cities (unitary authorities)</label>
-  <label><input type="radio" name="type" value="dc"
-      [% IF report_filter == 'dc' %]checked[% END %]> 
-      Districts and counties</label>
+<form method="GET">
+    <div class="filters">
+        <p>
+            [% IF area_body %]
+            <input type="hidden" name="area_body" value="[% area_body %]">
```
Comment (dracos) [inline templates/web/smidsy/reports/index.html]: It doesn't happen because area_body is used as an integer first, but note this is technically just relaying user inputted data, so should be escaped.

File: perllib/FixMyStreet/Cobrand/Smidsy.pm [hunk 9]
```diff
@@ -416,6 +416,307 @@ sub reports_hook_restrict_bodies_list {
...
+        $c->stash->{start_date} = DateTime::Format::Strptime->new( pattern => "%H:%M:%S %d:%m:%Y" )->format_datetime($start);
+        $c->stash->{end_date} = DateTime::Format::Strptime->new( pattern => "%H:%M:%S %d:%m:%Y" )->format_datetime($end);
+        $c->stash->{severities} = \@severities;
+        $c->stash->{reports_by_severity} = \%reports_by_severity;
+        $c->stash->{participants} = [ keys %reports_by_participants ];
+        $c->stash->{reports_by_participants} = \%reports_by_participants;
+        return 1;
+    }
+
+    return 0;
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Smidsy.pm]: This could be return 1 and no need for the two returns inside the if blocks.

File: perllib/FixMyStreet/Cobrand/Smidsy.pm [hunk 10]
```diff
@@ -416,6 +416,307 @@ sub reports_hook_restrict_bodies_list {
...
+        } else {
+            $area = $c->get_param('area_body') if $area eq 'wards';
+            $c->stash->{area_body} = $area;
+            $c->stash->{problems_rs} = $rs->search( { bodies_str => $area } );
+            $c->stash->{children} = $self->get_body_children( $area );
+        }
+    }
+
+    if ( $c->get_param('csv') ) {
+        $c->stash->{csv} = {
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Smidsy.pm]: Could this if and its else be split out into two functions?

File: perllib/FixMyStreet/Cobrand/Smidsy.pm [hunk 11]
```diff
@@ -416,6 +416,307 @@ sub reports_hook_restrict_bodies_list {
...
+            },
+            filename => 'collideoscope-data',
+        };
+        $self->generate_csv;
+        return 1;
+    } else {
+        $c->stash->{template} = 'reports/index.html';
+        unless ( $c->stash->{start_date} && $c->stash->{end_date} ) {
+            $c->forward('/reports/load_dashboard_data');
+            $c->stash->{problems_reported} = $c->stash->{problems_reported_by_period}->[-1];
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Smidsy.pm]: Any reason this is here and not in the template using `problems_reported = problems_reported_by_period.last | format_number` as in base?

File: perllib/FixMyStreet/Cobrand/Smidsy.pm [hunk 12]
```diff
@@ -416,6 +416,307 @@ sub reports_hook_restrict_bodies_list {
...
+                state => [ FixMyStreet::DB::Result::Problem->visible_states() ],
+            );
+
+            my @problems_reported_by_period;
+            foreach (map { $_->{n} } @problem_periods) {
+                push @problems_reported_by_period, ($problems_reported_by_period[-1]||0) + ($problems_reported_by_period{$_}||0);
+            }
+
+            $c->stash->{problem_periods} = [ map { $_->{d} || $_->{n} } @problem_periods ];
+            $c->stash->{problems_reported} = $problems_reported_by_period[-1];
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Smidsy.pm]: Ditto

File: perllib/FixMyStreet/Cobrand/Smidsy.pm [hunk 13]
```diff
@@ -416,6 +416,307 @@ sub reports_hook_restrict_bodies_list {
...
+            )->parse_datetime($c->stash->{start_date});
+            my $loop_start = $start->clone; # to avoid loop_period changing start
+
+            $end = DateTime::Format::Strptime->new(
+                pattern => '%F', # yyyy-mm-dd
+            )->parse_datetime($c->stash->{end_date});
+
+            use FixMyStreet::Script::UpdateAllReports;
+
+            my ($period, $extra);
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Smidsy.pm]: Again, feels like this should be factored with the identical/similar code in UpdateAllReports.

File: perllib/FixMyStreet/Cobrand/Smidsy.pm [hunk 14]
```diff
@@ -416,6 +416,307 @@ sub reports_hook_restrict_bodies_list {
...
+        $c->stash->{reports_by_severity} = \%reports_by_severity;
+        $c->stash->{participants} = [ keys %reports_by_participants ];
+        $c->stash->{reports_by_participants} = \%reports_by_participants;
+        return 1;
+    }
+
+    return 0;
+}
+
+sub generate_csv {
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Smidsy.pm]: Why is this here, why can't the copy in Dashboard be used?

File: perllib/FixMyStreet/Cobrand/Smidsy.pm [hunk 15]
```diff
@@ -416,6 +416,307 @@ sub reports_hook_restrict_bodies_list {
...
+                my $data = $report->get_extra_metadata;
+
+                return {
+                    external_body => $report->external_body,
+                    participants => $data->{participants} || '',
+                };
+            },
+            filename => 'collideoscope-data',
+        };
+        $self->generate_csv;
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Smidsy.pm]: forward to /dashboard/generate_csv?

File: templates/web/smidsy/reports/index.html [hunk 16]
```diff
@@ -1,55 +1,145 @@
...
+<div class="dashboard-row">
+    <div class="dashboard-item dashboard-item--6">
+        <h2 class="dashboard-subheading">[% loc('Download the data on this page') %]</h2>
+        <p>Grab a CSV of all the reports matching the filters you’ve set at the top of this page, suitable for analysis in applications like Excel or Fusion Tables.</p>
+        <a href="[% c.uri_with( {csv => 1} ) %]" class="btn">Download CSV</a>
+    </div>
+    <div class="dashboard-item dashboard-item--6">
+        <form class="dashboard-search">
+            <h2>[% loc('Show reports on a map') %]</h2>
+            <label for="body">[% loc('Pick a council') %]</label>
```
Comment (zarino) [inline templates/web/smidsy/reports/index.html]: Can we make this "Find" instead of "Pick"? As @LouiseMySociety noted, it’s a search box, you can’t "pick" anything until you’ve typed a search term.
